frontends/krita/krita_diff/script.py

```diff
@@ -357,11 +349,37 @@ def apply_img2img(self, is_inpaint):
             save_img(sel_image, path)
 
         def cb(response):
+            def cb_upscale(upscale_response):
+                if len(self.client.long_reqs) == 1:  # last request
+                    self.eta_timer.stop()
+                assert response is not None, "Backend Error, check terminal"
+
+                outputs = upscale_response["images"]
+                layer_name_prefix = "inpaint" if is_inpaint else "img2img"
+                glayer_name, layer_names = get_desc_from_resp(response, layer_name_prefix)
+                layers = [
+                    insert(name if name else f"{layer_name_prefix} {i + 1}", output)
+                    for output, name, i in zip(outputs, layer_names, itertools.count())
+                ]
+                if self.cfg("hide_layers", bool):
+                    for layer in layers[:-1]:
+                        layer.setVisible(False)
+                if glayer:
+                    glayer.setName(glayer_name)
+                self.doc.refreshProjection()
+                self.controlnet_transparency_mask_inserter(layers, mask_image)
```

Comment:
Check if it's inpaint before working with the transparency layer (got an error on my end).
```suggestion
                if is_inpaint:
                    self.controlnet_transparency_mask_inserter(layers, mask_image)
```

Reply:
The original transparency mask inserter I based that on actually operated on img2img exclusively (where it would add a mask based on the current selection), so it should be correct functionality for it to operate on both img2img and inpainting.  I do need to do a pass over img2img support, though.

I'm converting this to a draft in the meantime -- I have to set it up to skip the upscaler API if it is not needed, for simplicity I might just pass back a fake API response through the callback.  I have also found that the race conditions are still present particularly on lower end machines, but I'm having trouble reproducing them on my own machine.  I may opt for the nuclear option and adopt a single approach for handling transparency masks -- always have layer groups, have a transparency mask for the layer group, and then add the outputs.  The layer group is the only one that stores generation parameters anyways, and this is more efficient than having four identical transparency masks.  The result will be the same -- you will have a group with a layer for each result in your batch, you will pick one to make visible, and you can then flatten the group.